### ISSUE TITLE
RFC on parallelizing --brute

### ISSUE THREAD

**Issue by apankrat** [upx/upx#310]
## What's the problem (or question)?

Can I ask if there exists an opinion on making --brute multi-threaded?

I'm sitting on 16 cores. It's a shame that 15 of them don't get to participate.

Having skimmed through packer.cpp code, it looks like that two main `for` loops can be reworked to queue tasks for a pool of "worker" thread and then digest the results. In theory. In practice, of course, this will require a non-trivial rework and we all know how much fun it is.

Hence the question - what's the general view on this? "Heck no", "Why not", "You do it", etc.

## What should have happened?

n/a

## Do you have an idea for a solution?

I can give it a go, just wanted to check that if it's been considered or if there are any fundamental objections to doing this.

## How can we reproduce the issue?

n/a

## Please tell us details about your environment.

Lots of unused CPU cores.

**Comment by jreiser** [comment]
The general view is that parallelizing upx is not worth the complexity.  Remember that upx desires to totally cross-platform: while running ON any platform, then upx compresses executables FOR any platform.  Getting multi-threading to work on a wide variety of platforms is tedious.  There are bigger fish to fry, such as Android (runtime dynamic linking is VERY quirky), Windows10 (PE format sprouted weeds), etc.
To compress multiple executables, then parallelize at shell level: invoke upx as many times as there are executables to compress.  So parallelizing upx is for using multiple threads on a single input file.  If the file is larger than about 30KB, then always --lzma will be the smallest encoding, and only the filter and the search space for deflation matter.  In practice the best filter for a given input file never changes.  The search space for --brute is quadratic, and threads offer at most a constant factor reduction.

**Comment by apankrat** [comment]
> Getting multi-threading to work on a wide variety of platforms is tedious.

No argument there, but that would've not been the goal.

Basically I was thinking to split the for-for search loop into a prep part (to generate a "todo" list) and a part to execute this list. Hide the guts of the latter behind some simple abstraction and default to a dumb loop. Then, for platforms, where it's feasible, replace the loop with a threaded version.

> If the file is larger than about 30KB, then always --lzma will be the smallest encoding

Good to know. Production releases don't report the winning combination though, which ties into...

> In practice the best filter for a given input file never changes.

It'd be _very_ handy if a `--brute` run would report what the winning combo was, so that it could be fed into a non-brute invocation. Ideally as a simple string (or hex number), so that it'd be a bit more script-friendly. If it's OK in principle, I can take a crack at this.

**Comment by jreiser** [comment]
```
$ ./upx.out --fileinfo foo
foo [amd64-linux.elf, linux/amd64]
     53368 bytes, compressed by UPX 13, method 2, level 8, filter 0x49/0x08
```
method=2 ==> conf.h: #define M_NRV2B_LE32    2
level=8 ==> 1..9
filter 0x49/08 ==> filteri.cpp:    { 0x49, 6, 0x00ffffff, f_ctok32_e8e9_bswap_le, u_ctok32_e8e9_bswap_le, s_ctok32_e8e9_bswap_le },

**Comment by apankrat** [comment]
Most excellent, thanks for the pointer.

I did manage to built `upx.exe` using stock Visual Studio 2017 just now, sprinkled some printfs and poked around a bit. This made me realize that I ended up looking at `--brute` for the wrong reasons. 

See, we pack the binary with `--lzma`. Today I re-tested it with `--brute` and the result was noticeably smaller, but since I didn't know how to see the resulting method/filter combination I *assumed* it must be some combo well off the beaten path. Which is why I thought we should start doing `--brute` for every binary => it had to be fast => a question about parallel processing.

But now, having looked at printf traces, it's obvious that the difference was due to `--brute` automatically defaulting to the `--best` level instead of `-7` for the non-brute invocation. The obvious fix is just to add `--best` to our packaging script.

Case closed, but I think it'd be helpful for people like myself to have your *very* valuable "30KB / LZMA" heuristic and an advise to use `--best` mentioned somewhere in the README if looking for the smallest size.
